cmd/syncthing/gui.go

```diff
@@ -334,29 +338,33 @@ func (s *apiService) Serve() {
 	// Add the CORS handling
 	handler = corsMiddleware(handler, guiCfg.InsecureAllowFrameLoading)
 
-	if addressIsLocalhost(guiCfg.Address()) && !guiCfg.InsecureSkipHostCheck {
-		// Verify source host
-		handler = localhostMiddleware(handler)
+	if !guiCfg.InsecureSkipHostCheck {
+		hasRemoteAddr := false
+		for _, addr := range guiCfg.Addresses() {
+			if !addressIsLocalhost(addr) {
+				hasRemoteAddr = true
+				break
+			}
+		}
+		// Verify source host if all listening addresses are local.
+		if !hasRemoteAddr {
+			handler = localhostMiddleware(handler)
+		}
 	}
 
 	handler = debugMiddleware(handler)
 
-	srv := http.Server{
-		Handler: handler,
-		// ReadTimeout must be longer than SyncthingController $scope.refresh
-		// interval to avoid HTTP keepalive/GUI refresh race.
-		ReadTimeout: 15 * time.Second,
-	}
-
 	s.fss = newFolderSummaryService(s.cfg, s.model)
 	defer s.fss.Stop()
 	s.fss.ServeBackground()
 
-	l.Infoln("GUI and API listening on", listener.Addr())
-	l.Infoln("Access the GUI via the following URL:", guiCfg.URL())
+	for _, listener := range listeners {
+		l.Infoln("GUI and API listening on", listener.Addr())
+		l.Infoln("Access the GUI via the following URL:", guiCfg.UrlFromAddress(listener.Addr().String()))
+	}
 	if s.started != nil {
 		// only set when run by the tests
-		s.started <- listener.Addr().String()
+		s.started <- listeners[0].Addr().String()
```

Comment:
Indexing into a potentially empty array feels a bit scary.

Reply:
Maybe I am wrong but the listeners are obtained from the `GUIConfiguration` structure which specifies a default listener value (to `127.0.0.1:8384`).

Reply:
I guess, but I'd write a test to verify this, as in ensure that loading a config with no address section we still end up with atleast a single address entry.

This will furthermore need UI improvements, as in the UI currently listen address is a single field.

Reply:
Same comment as above, opening unix://foo/bar is not very useful, so you'd want to make sure you pick something that is http.